### PR TITLE
Move mixin to core

### PR DESCRIPTION
**Pull request by kaushikb11**
## What does this PR do?

Fixes #<issue_number>

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)
- [ ] Did you list all the breaking changes introduced by this pull request?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/core/datamodule.py [hunk 1]
```diff
@@ -20,9 +20,9 @@
 from torch.utils.data import DataLoader, Dataset, IterableDataset
 
 from pytorch_lightning.core.hooks import CheckpointHooks, DataHooks
+from pytorch_lightning.core.mixin import HyperparametersMixin
```
Comment (carmocca) [inline pytorch_lightning/core/datamodule.py:23]: Most directories are plural

```suggestion
from pytorch_lightning.core.mixins import HyperparametersMixin
```

**Comment by Borda** [comment]
just curious, what is the motivation?

File: pytorch_lightning/core/datamodule.py (hunk 1 above)
Comment (kaushikb11) [reply pytorch_lightning/core/datamodule.py]: Updated!

**Comment by kaushikb11** [comment]
@Borda [Reference](https://github.com/PyTorchLightning/pytorch-lightning/pull/3792#discussion_r666543813)

**Review by awaelchli — APPROVED** [review]
doesn't the device_dtype_mixin require a rerouting of the imports for BC?

**Comment by kaushikb11** [comment]
> doesn't the device_dtype_mixin require a rerouting of the imports for BC?

I don't think it would be required for internal use.

**Review by justusschock — APPROVED** [review]

**Comment by justusschock** [comment]
@kaushikb11 > I don't think it would be required for internal use.

Not for internal use, but it was part of the public API in utilities (I use it quite often in my own code) so I think we should reroute this together with a Deprecation Warning

**Review by tchaton — APPROVED** [review]
LGTM !
